File: go/test/volume_test.go
package test

import (
	"testing"

	modal "github.com/modal-labs/modal-client/go"
	"github.com/modal-labs/modal-client/go/internal/grpcmock"
	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"github.com/onsi/gomega"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
)

func TestVolumeFromName(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	volume, err := tc.Volumes.FromName(ctx, "libmodal-test-volume", &modal.VolumeFromNameParams{
		CreateIfMissing: true,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(volume).ShouldNot(gomega.BeNil())
	g.Expect(volume.VolumeID).Should(gomega.HavePrefix("vo-"))
	g.Expect(volume.Name).To(gomega.Equal("libmodal-test-volume"))

	_, err = tc.Volumes.FromName(ctx, "missing-volume", nil)
	g.Expect(err).Should(gomega.MatchError(gomega.ContainSubstring("Volume 'missing-volume' not found")))
}

func TestVolumeReadOnly(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	volume, err := tc.Volumes.FromName(ctx, "libmodal-test-volume", &modal.VolumeFromNameParams{
		CreateIfMissing: true,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(volume.IsReadOnly()).To(gomega.BeFalse())

	readOnlyVolume := volume.ReadOnly()
	g.Expect(readOnlyVolume.IsReadOnly()).To(gomega.BeTrue())
	g.Expect(readOnlyVolume.VolumeID).To(gomega.Equal(volume.VolumeID))
	g.Expect(readOnlyVolume.Name).To(gomega.Equal(volume.Name))

	g.Expect(volume.IsReadOnly()).To(gomega.BeFalse())
}

func TestVolumeEphemeral(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	tc := newTestClient(t)

	volume, err := tc.Volumes.Ephemeral(t.Context(), nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer volume.CloseEphemeral()
	g.Expect(volume.Name).To(gomega.BeEmpty())
	g.Expect(volume.VolumeID).Should(gomega.HavePrefix("vo-"))
	g.Expect(volume.IsReadOnly()).To(gomega.BeFalse())
	g.Expect(volume.ReadOnly().IsReadOnly()).To(gomega.BeTrue())
}

func TestVolumeDeleteSuccess(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/VolumeGetOrCreate",
		func(req *pb.VolumeGetOrCreateRequest) (*pb.VolumeGetOrCreateResponse, error) {
			return pb.VolumeGetOrCreateResponse_builder{
				VolumeId: "vo-test-123",
			}.Build(), nil
		},
	)

	grpcmock.HandleUnary(
		mock, "/VolumeDelete",
		func(req *pb.VolumeDeleteRequest) (*emptypb.Empty, error) {
			g.Expect(req.GetVolumeId()).To(gomega.Equal("vo-test-123"))
			return &emptypb.Empty{}, nil
		},
	)

	err := mock.Volumes.Delete(ctx, "test-volume", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestVolumeDeleteWithAllowMissing(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/VolumeGetOrCreate",
		func(req *pb.VolumeGetOrCreateRequest) (*pb.VolumeGetOrCreateResponse, error) {
			return nil, modal.NotFoundError{Exception: "Volume 'missing' not found"}
		},
	)

	err := mock.Volumes.Delete(ctx, "missing", &modal.VolumeDeleteParams{
		AllowMissing: true,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestVolumeDeleteWithAllowMissingDeleteRPCNotFound(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(mock, "/VolumeGetOrCreate",
		func(req *pb.VolumeGetOrCreateRequest) (*pb.VolumeGetOrCreateResponse, error) {
			return pb.VolumeGetOrCreateResponse_builder{VolumeId: "vo-test-123"}.Build(), nil
		},
	)

	grpcmock.HandleUnary(mock, "/VolumeDelete",
		func(req *pb.VolumeDeleteRequest) (*emptypb.Empty, error) {
			return nil, status.Errorf(codes.NotFound, "Volume not found")
		},
	)

	err := mock.Volumes.Delete(ctx, "test-volume", &modal.VolumeDeleteParams{AllowMissing: true})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestVolumeDeleteWithAllowMissingFalseThrows(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/VolumeGetOrCreate",
		func(req *pb.VolumeGetOrCreateRequest) (*pb.VolumeGetOrCreateResponse, error) {
			return nil, modal.NotFoundError{Exception: "Volume 'missing' not found"}
		},
	)

	err := mock.Volumes.Delete(ctx, "missing", &modal.VolumeDeleteParams{
		AllowMissing: false,
	})
	g.Expect(err).Should(gomega.HaveOccurred())
	var notFoundErr modal.NotFoundError
	g.Expect(err).Should(gomega.BeAssignableToTypeOf(notFoundErr))

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

File: js/test/volume.test.ts
import { tc } from "../test-support/test-client";
import { expect, onTestFinished, test } from "vitest";
import { createMockModalClients } from "../test-support/grpc_mock";
import { NotFoundError } from "../src/errors";
import { volumeToMountProto } from "../src/volume";
import { ClientError, Status } from "nice-grpc";

test("client.volumes.fromName", async () => {
  const volume = await tc.volumes.fromName("libmodal-test-volume", {
    createIfMissing: true,
  });
  expect(volume.volumeId).toMatch(/^vo-/);
  expect(volume.name).toBe("libmodal-test-volume");

  const promise = tc.volumes.fromName("missing-volume");
  await expect(promise).rejects.toThrowError(
    /Volume 'missing-volume' not found/,
  );
});

test("Volume.withMountOptions", async () => {
  const volume = await tc.volumes.fromName("libmodal-test-volume", {
    createIfMissing: true,
  });

  const mount = volume.withMountOptions({
    readOnly: true,
    subPath: "/items",
  });
  expect(mount.volumeId).toBe(volume.volumeId);

  const mountProto = volumeToMountProto("/mnt", mount);
  expect(mountProto.readOnly).toBe(true);
  expect(mountProto.subPath).toBe("/items");

  const unconfiguredProto = volumeToMountProto("/mnt", volume);
  expect(unconfiguredProto.readOnly).toBe(false);
  expect(unconfiguredProto.subPath).toBeUndefined();
});

test("Volume.withMountOptions stacks", async () => {
  const volume = await tc.volumes.fromName("libmodal-test-volume", {
    createIfMissing: true,
  });

  const configured = volume.withMountOptions({
    readOnly: true,
    subPath: "/nested",
  });

  // Setting only subPath preserves readOnly from the previous call.
  const withNewSubPath = configured.withMountOptions({ subPath: "/other" });
  const newSubPathProto = volumeToMountProto("/mnt", withNewSubPath);
  expect(newSubPathProto.readOnly).toBe(true);
  expect(newSubPathProto.subPath).toBe("/other");

  // Setting only readOnly preserves subPath from the previous call.
  const withReadOnlyDisabled = configured.withMountOptions({ readOnly: false });
  const readOnlyDisabledProto = volumeToMountProto(
    "/mnt",
    withReadOnlyDisabled,
  );
  expect(readOnlyDisabledProto.readOnly).toBe(false);
  expect(readOnlyDisabledProto.subPath).toBe("/nested");

  // subPath "/" is normalized to undefined (mount the whole volume).
  const withClearedSubPath = configured.withMountOptions({ subPath: "/" });
  const clearedSubPathProto = volumeToMountProto("/mnt", withClearedSubPath);
  expect(clearedSubPathProto.readOnly).toBe(true);
  expect(clearedSubPathProto.subPath).toBeUndefined();
});

test("VolumeEphemeral", async () => {
  const volume = await tc.volumes.ephemeral();
  onTestFinished(() => volume.closeEphemeral());

  expect(volume.name).toBeUndefined();
  expect(volume.volumeId).toMatch(/^vo-/);
  const readOnlyVolume = volume.withMountOptions({ readOnly: true });
  expect(volumeToMountProto("/mnt", readOnlyVolume).readOnly).toBe(true);
});

test("VolumeDelete success", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/VolumeGetOrCreate", () => ({
    volumeId: "vo-test-123",
    metadata: { name: "test-volume" },
  }));

  mock.handleUnary("/VolumeDelete", (req: any) => {
    expect(req.volumeId).toBe("vo-test-123");
    return {};
  });

  await mc.volumes.delete("test-volume");
  mock.assertExhausted();
});

test("VolumeDelete with allowMissing=true", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/VolumeGetOrCreate", () => {
    throw new NotFoundError("Volume 'missing' not found");
  });

  await mc.volumes.delete("missing", { allowMissing: true });
  mock.assertExhausted();
});

test("VolumeDelete with allowMissing=true when delete RPC returns NOT_FOUND", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  // fromName succeeds — volume exists at lookup time
  mock.handleUnary("/VolumeGetOrCreate", () => ({
    volumeId: "vo-test-123",
    metadata: { name: "test-volume" },
  }));

  // volumeDelete fails — volume was deleted between lookup and delete
  mock.handleUnary("/VolumeDelete", () => {
    throw new ClientError(
      "/modal.client.ModalClient/VolumeDelete",
      Status.NOT_FOUND,
      "No Volume with ID 'vo-test-123' found",
    );
  });

  // With allowMissing=true, this should succeed silently
  await mc.volumes.delete("test-volume", { allowMissing: true });
  mock.assertExhausted();
});

test("VolumeDelete with allowMissing=false throws", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/VolumeGetOrCreate", () => {
    throw new NotFoundError("Volume 'missing' not found");
  });

  await expect(
    mc.volumes.delete("missing", { allowMissing: false }),
  ).rejects.toThrow(NotFoundError);
});
